### ISSUE TITLE
Import an IFC using the angle to true north to influence rotation

### ISSUE THREAD

**Issue by Moult** [Autodesk/revit-ifc#321]
When an IFC is imported, its local coordinates are always lined up to project north. This is a problem, because some IFC files (i.e. from ArchiCAD) are already rotated such that grid north is orthogonal, not project north.

In Revit, there seems to be no way to influence the rotation of the imported IFC.

This has been mentioned before from web searches but with no solution:

 - https://forums.autodesk.com/t5/revit-structure-forum/export-link-ifc/td-p/5817251
 - https://forums.autodesk.com/t5/revit-architecture-forum/linking-ifc-rvt-files-into-projects-with-project-north-differing/td-p/8206106

Thoughts?

**Comment by DuncanLithgow** [comment]
@Moult when you say 'grid north' I assume you mean compass / map / true north right?

**Comment by Moult** [comment]
@DuncanLithgow actually, I do mean grid north. There are a few norths out there:

 - Project north, an arbitrary orientation usually decided by the architect so that your building is drawn orthogonal
 - Grid north, a direction towards north along the grid lines of a map projection. So when your surveyor gives you a survey with typically big numbers for eastings and northings, they use a map projection system (in Sydney we might use GDA94 / MGA 56, a.k.a. EPSG:28356). So when you georeference your model, "grid north" is up the page. This grid is typically orthogonal.
 - True north, a.k.a geodetic north, the direction to the geographic north pole, the northenmost point of the earth. This direction is along the latitudes of the Earth. This is not the same as grid north. It typically deviates quite a bit. Since surveyors draw using grid north, this means that true north is actually a long a curve (i.e. along a latitude) and is not a constant. This true north is what we should use for solar / shadow analysis, and I've sometimes heard it called solar north.
 - Magnetic north, which likes to shift around and is typically a bit useless for the AEC industry.

So, we draw in project north, survey / georeference in grid north, and solar analysis in true north.

Out of the three important ones, Revit supports only 2 depending on how you use Revit. This discrepancy leads to incorrect georeferencing or incorrect solar analysis typically. However, especially now that Revit is actually implementing IFC map conversion, getting grid north correct has the higher priority. Fixing Revit to work with the three norths is another issue. There should probably be another bug filed for how Revit stores true north in IFC/

**Comment by eibre** [comment]
Has this changed recently? When I link an IFC now the coordinates seems to align origin to origin but the rotation is alignet with the current "true north" angle

![image](https://user-images.githubusercontent.com/5968613/207816697-2d8552c5-84e7-476b-9f64-bb6041dd7d98.png)

**Comment by Moult** [comment]
@eibre yes there was a change: https://github.com/Autodesk/revit-ifc/issues/522

**Comment by thumDer** [comment]
There should be multiple options on the linking dialog, including the coordinate mapping. Now it is an inconsistent guessing game, which never behaves as expected.

**Comment by ArtemHnatko** [comment]
Thank you for the detailed report and for explaining the grid north / true north / project north distinction so clearly — it's a helpful framing for this whole class of issues.

After investigating the current codebase and related issues, we believe this is related to **#522** ([Revit should not link IFCs and make project north = grid north](https://github.com/Autodesk/revit-ifc/issues/522)), which is tracked internally as **REVIT-245566**.

The root cause is in the same area: the importer (`IFCProject.cs`, `UpdateProjectLocation`) treats the `IfcMapConversion` X-axis rotation — which the IFC schema defines as **grid north** — as Revit's **true north**, with no user control over this behavior. The issue is confirmed present through Revit 2027 / IFC exporter v27.0.1.1.

Related open issues in this cluster:
- [#522](https://github.com/Autodesk/revit-ifc/issues/522) — Revit should not link IFCs and make project north = grid north
- [#710](https://github.com/Autodesk/revit-ifc/issues/710) — Different project north rotations: linking IFC vs `.ifc.rvt`
- [#465](https://github.com/Autodesk/revit-ifc/issues/465) — Revit doesn't differentiate between grid north and true north
- [#16](https://github.com/Autodesk/revit-ifc/issues/16) — Link IFC needs options for placement and rotation

**Comment by Moult** [comment]
I think this is a separate, but related bug:

This bug: Revit hardcodes imports such that project north is up (sorta, but the point is that it is fixed, see #710). In reality, some projects will have different project norths and this is currently ignored. This is about calculating the deviation in project north and handling it appropriately.

Bug #522: Revit imports IFC's project north (+Y) to point to Revit's true north (which is wrong). This was changed behaviour in Revit between 2020 and 2022 (not at computer to verify current behaviour).

BTW, for georef and coordinates, it would perhaps be beneficial if we organised a video call / screenshare to go through all the scenarios and testcases, and I can provide test case IFC files too.

**Comment by ArtemHnatko** [comment]
Hello @Moult,

Thank you for the clarification — and I completely agree with your distinction.

You are absolutely right that these are separate, though closely related, issues:

#321 deals with the absence of user control over rotation upon import and the incorrect hardcoding of "project north is up," which fails when the source IFC uses a different project north convention (e.g., ArchiCAD exports). This is indeed a broader design limitation.

#522 is a specific regression (introduced between Revit 2020 and 2022) where the IFC’s Project North (+Y) is incorrectly mapped to the host’s True North, rather than remaining aligned with Project North.

When I marked #321 as a duplicate, my intent was to consolidate discussion around the root cause — the flawed logic in IFCProject.cs / UpdateProjectLocation that mishandles north rotations in general. However, I fully acknowledge that from a user and testing perspective, these are two distinct symptoms that may require separate fixes or at least separate test cases.

I will update the ticket status to reflect that #321 is related to rather than strictly "identical to" #522, and we will track both scenarios under the same internal issue (REVIT-245566) with clear sub-cases.

A video call / screen share to walk through the different north scenarios and coordinate workflows sounds useful. Having IFC test files covering these cases would also help us validate the expected behavior more accurately.

Thank you for providing the additional context and examples.
